src/Esprima/JavascriptParser.cs

```diff
@@ -3877,17 +3877,20 @@ private RestElement ParseRestElement(ref ArrayList<Token> parameters)
         return Finalize(node, new RestElement(arg));
     }
 
+    // pooled for ParseLexicalBinding calls
+    private ArrayList<Token> _parseFormalParameterParameters;
+
     private void ParseFormalParameter(ref ParsedParameters options)
     {
-        var parameters = new ArrayList<Token>();
+        _parseFormalParameterParameters.Clear();
```

Comment:
Consider the following input: `function f(x, y = function(a, b) {}, z) {}`

In such cases `ParseFormalParameter` is re-entered. So, without saving the current `_parseFormalParameterParameters` into a local variable and restoring it at the end of the method the pooling would be buggy.

The same goes for `ParseVariableDeclaration` and `ParseLexicalBinding`. (BTW, `ParseCatchClause` and `ParseGroupExpression` are candidates for pooling too).

A further remark: do we really need multiple `ArrayList<Token>` fields (`_parseLexicalBindingParameters`, `_parseFormalParameterParameters`, etc.) for this kind of pooling? A single one looks sufficient me (after ironing out the aforementioned problem, of course).